frb_codegen/src/ir/file.rs

```diff
@@ -1,29 +1,123 @@
+use std::thread;
+
+use crate::utils::misc::ExtraTraitForVec;
 
 pub type IrStructPool = HashMap<String, IrStruct>;
 pub type IrEnumPool = HashMap<String, IrEnum>;
 
+use std::cell::RefCell;
+
+// this variable should be `none` for single block case, and not none for multi-blocks case.
+thread_local!(pub static SHARED_MODULE: RefCell<Option<String>> = RefCell::new(None));
+
+// shared types among all regular API blocks, in context of function paramters
+thread_local!(pub static SHARED_TYPES_INPUT: RefCell<HashSet<IrType>> = RefCell::new(HashSet::new()));
+thread_local!(static FETCHED_FOR_SHARED_TYPES_INPUT: RefCell<bool> = RefCell::new(false));
+
+// shared types among all regular API blocks in context of function outputs
+thread_local!(static SHARED_TYPES_OUTPUT: RefCell<HashSet<IrType>> = RefCell::new(HashSet::new()));
+thread_local!(static FETCHED_FOR_SHARED_TYPES_OUTPUT: RefCell<bool> = RefCell::new(false));
+
+// shared types among all regular API blocks
+thread_local!(static SHARED_TYPES_ALL: RefCell<HashSet<IrType>> = RefCell::new(HashSet::new()));
+thread_local!(static FETCHED_FOR_SHARED_TYPES_ALL: RefCell<bool> = RefCell::new(false));
+
 #[derive(Debug, Clone)]
 #[cfg_attr(feature = "serde", derive(serde::Serialize))]
 pub struct IrFile {
-    pub funcs: Vec<IrFunc>,
+    /// for a regular API block: this field contains APIs defined in this block and
+    /// all methods of types(struct/enum) used ONLY in this block
+    /// ---NOTE: for shared types used in this block, their methods are also included here;
+    /// for a shared block: it ONLY contains methods from ALL shared types.
+    funcs: Vec<IrFunc>,
     pub struct_pool: IrStructPool,
     pub enum_pool: IrEnumPool,
     pub has_executor: bool,
+    pub block_index: BlockIndex,
+    pub shared: bool, // `true` for an auto-generated shared block, otherwise `false`
 }
 
 impl IrFile {
+    pub fn funcs(&self, exclude_shared_method: bool) -> Vec<IrFunc> {
+        if self.shared {
+            // assert each item of funcs are shared
+            for each in &self.funcs {
+                assert!(each.shared);
+            }
+            return self.funcs.clone();
+        }
+        match exclude_shared_method {
+            true => self
+                .funcs
+                .iter()
+                .cloned()
+                .filter(|each| !each.shared)
+                .collect::<Vec<_>>(),
+            false => self.funcs.clone(),
+        }
+    }
+
+    pub fn set_funcs(&mut self, dst_funcs: Vec<IrFunc>) {
+        self.funcs = dst_funcs;
+    }
+
+    pub fn append_funcs(&mut self, dst_funcs: Vec<IrFunc>) {
+        self.funcs.extend(dst_funcs);
+    }
+
+    /// Returns a vector of all shared methods in the file.
+    /// For a regular API block, this method returns all methods of shared types used in this block.
+    /// For a shared block, it returns all methods from all shared types.
+    pub fn get_shared_methods(&self) -> Vec<IrFunc> {
+        if !self.shared {
```

Comment:
if !sth -> if sth

Reply:
For this, I made `   if !self.shared {` as the 1st condition on purpose, because I want to check the regular block first, since it is the most common case. Thus, there are lots of `   if !self.shared {` in this pr.  And I do know this would somehow make it anti-intuitionally think that the 1st block is for `shared`.  So do you want I change all the similar case to  `if shared` first?

Reply:
I realize that, if we change the `shared:bool` to an enum (`enum BlockMode {shared block, regular block}`), as is mentioned in another comment, then we will not be faced with this problem indeed.

Reply:
There are 2 cases related to the `shared:bool` flag. One is for the whole block, like that for struct `IrFile`, another is for types(struct,enum...). So, what if just using 1, say... `enum ShareMode {unique, shared}`?

Reply:
LGTM with a brief look, as long as the semantics is reasonable!

Reply:
Refined in [4322b5d](https://github.com/fzyzcjy/flutter_rust_bridge/pull/1145/commits/4322b5d7231852ffd673df16ba155433afbf9f89).      

Note: When using the new `ShareMode` enum, I insist on using `match` instead of `if ...{} else {}` for pair checking in most cases. This is because custom enum types may have additional items added to them in the future, and using `match` ensures that we developers are forced to check for new items when they are added. On the other hand, using `if ...{} else {}` may leave potential pitfalls if new items are added and not accounted for.